src/realm/sync/noinst/migration_store.cpp

```diff
@@ -0,0 +1,126 @@
+#include <realm/sync/noinst/migration_store.hpp>
+
+#include <realm/transaction.hpp>
+#include <realm/sync/noinst/sync_metadata_schema.hpp>
+
+namespace realm::sync {
+namespace {
+constexpr static int c_schema_version = 1;
+constexpr static std::string_view c_flx_migration_table("flx_migration");
+constexpr static std::string_view c_flx_migration_started_at("flx_migration_started_at");
+constexpr static std::string_view c_flx_migration_completed_at("flx_migration_completed_at");
+constexpr static std::string_view c_flx_migration_state("flx_migration_state");
+constexpr static std::string_view c_flx_migration_query_string("flx_migration_query_string");
+
+constexpr static std::string_view c_flx_subscription_name_prefix("flx_migrated_");
+
+class MigrationStoreInit : public MigrationStore {
+public:
+    explicit MigrationStoreInit(DBRef db,
+                                std::function<void(MigrationStore::MigrationState)>&& on_migration_state_changed)
+        : MigrationStore(std::move(db), std::move(on_migration_state_changed))
+    {
+    }
+};
+
+} // namespace
+
+MigrationStoreRef
+MigrationStore::create(DBRef db, std::function<void(MigrationStore::MigrationState)>&& on_migration_state_changed)
+{
+    return std::make_shared<MigrationStoreInit>(std::move(db), std::move(on_migration_state_changed));
+}
+
+MigrationStore::MigrationStore(DBRef db,
+                               std::function<void(MigrationStore::MigrationState)>&& on_migration_state_changed)
+    : m_db(std::move(db))
+    , m_on_migration_state_changed(std::move(on_migration_state_changed))
+{
+    std::vector<SyncMetadataTable> internal_tables{
+        {&m_migration_table,
+         c_flx_migration_table,
+         {
+             {&m_migration_started_at, c_flx_migration_started_at, type_Timestamp},
+             {&m_migration_started_at, c_flx_migration_completed_at, type_Timestamp},
+             {&m_migration_state, c_flx_migration_state, type_Int},
+             {&m_migration_query_str, c_flx_migration_query_string, type_String},
+         }},
+    };
+
+    auto tr = m_db->start_read();
+    SyncMetadataSchemaVersions schema_versions(tr);
+
+    if (auto schema_version = schema_versions.get_version_for(tr, internal_schema_groups::c_flx_migration_store);
+        !schema_version) {
+        tr->promote_to_write();
+        schema_versions.set_version_for(tr, internal_schema_groups::c_flx_migration_store, c_schema_version);
+        create_sync_metadata_schema(tr, &internal_tables);
+        // create migration object
+        auto migration_store_obj = tr->get_table(m_migration_table)->create_object();
+        migration_store_obj.set(m_migration_started_at, Timestamp{std::chrono::system_clock::now()});
+        migration_store_obj.set(m_migration_state, MigrationState::InProgress);
```

Comment:
The `migrate_to_flx()` function will update the state to InProgress/Migrated. When this object is created, it should just read the current metadata values, if any, or default to not migrated if the metadata table doesn't exist.